File: expertAgent/aiagent/langgraph/jobGeneratorV2/workflows/workflow_gen/prompt_builder/rules/taskflow_rules.py
"""TaskFlow V2 Prompt Rules.

Issue #350 Task 2.2: TaskFlow-specific LLM prompt rules.
Issue #350 Fix: Added API capability rules for expertAgent internal APIs.

This module provides prompt rules for generating TaskFlow V2 JSON workflows.
"""

from __future__ import annotations

from core.config import settings


def get_taskflow_rules() -> list[str]:
    """Get TaskFlow V2 generation rules.

    Returns:
        List of rule sections for LLM prompts
    """
    return [
        get_taskflow_structure_rules(),
        get_taskflow_step_types(),
        get_taskflow_variable_reference_rules(),
        get_taskflow_security_rules(),
        get_taskflow_api_rules(),
    ]


def get_taskflow_structure_rules() -> str:
    """Get TaskFlow V2 structure rules."""
    return """## TaskFlow V2 Workflow Structure

A TaskFlow V2 workflow is a JSON document with the following structure:

```json
{
  "workflow_name": "string (required, pattern: ^[a-zA-Z_][a-zA-Z0-9_-]*$)",
  "description": "string (optional)",
  "input_schema": {"field_name": "type"},
  "output_schema": {"field_name": "type"},
  "steps": [/* step definitions */],
  "output": {"field_name": "${step_id.output}"}
}
```

### Field Types
- string, number, boolean, array, object, null

### Step ID Pattern
Step IDs must match: `^[a-zA-Z_][a-zA-Z0-9_-]*$`
- Valid: `fetch_data`, `step_001`, `process_result`
- Invalid: `123step`, `step@name`, `step name`
"""


def get_taskflow_step_types() -> str:
    """Get TaskFlow V2 step type descriptions."""
    return """## TaskFlow V2 Step Types

### 1. api_rest - REST API Call
```json
{
  "id": "fetch_user",
  "type": "api_rest",
  "config": {
    "step_type": "api_rest",
    "method": "GET|POST|PUT|DELETE|PATCH",
    "url": "https://api.example.com/endpoint",
    "headers": {"Authorization": "Bearer ${secrets.TOKEN}"},
    "body": "{\"key\": \"${inputs.value}\"}",
    "timeout_ms": 30000
  }
}
```
**IMPORTANT**:
- URL must use HTTPS. HTTP is not allowed.
- config must include "step_type" matching the step type.
- **body must be a JSON string**, not a dict object. Use escaped quotes.

### 2. transform - Data Transformation

**CRITICAL**: Each transform mode requires SPECIFIC config fields. DO NOT use fields not listed for each mode.

#### Mode: template (most common)
Use `template` field with ${} variable references:
```json
{
  "id": "format_output",
  "type": "transform",
  "config": {
    "step_type": "transform",
    "mode": "template",
    "template": "Result: ${fetch_user.output.data.name}"
  }
}
```

**CRITICAL**: The `template` field MUST be a STRING, not a dict/object!
- ✅ CORRECT: `"template": "Result: ${step.output.data}"`
- ✅ CORRECT: `"template": "${step.output.data.name}"`
- ❌ WRONG: `"template": {"result": "${step.output.data}"}`  ← This is INVALID!
- ❌ WRONG: `"template": {"search_results": "${api.output.data.items}"}`  ← This is INVALID!

If you need structured output, use a simple string template and let the output mapping handle it:
```json
{
  "id": "format_output",
  "type": "transform",
  "config": {
    "step_type": "transform",
    "mode": "template",
    "template": "${api_call.output.data}"
  }
}
```
Then use the workflow output field to structure the results:
```json
"output": {
  "result": "${format_output.output}",
  "status": "${api_call.output.status}"
}
```

#### Mode: concat
Use `separator` field to join array elements:
```json
{
  "id": "join_items",
  "type": "transform",
  "config": {
    "step_type": "transform",
    "mode": "concat",
    "separator": ", "
  }
}
```

#### Mode: map
Use `fields` field (array of field names to extract):
```json
{
  "id": "extract_fields",
  "type": "transform",
  "config": {
    "step_type": "transform",
    "mode": "map",
    "fields": ["name", "email", "id"]
  }
}
```
**WARNING**: DO NOT use `map: {...}` - this is WRONG. Use `fields: [...]` instead.

#### Mode: merge
Use `fields` field (array of field names to merge):
```json
{
  "id": "merge_data",
  "type": "transform",
  "config": {
    "step_type": "transform",
    "mode": "merge",
    "fields": ["user", "profile"]
  }
}
```

**RECOMMENDATION**: For most data transformation needs, use `mode: "template"` with ${} references.

### 3. code_js - JavaScript Function (Limited)
```json
{
  "id": "parse_data",
  "type": "code_js",
  "config": {
    "step_type": "code_js",
    "path": "/scripts/utils.js",
    "function_name": "parseJson"
  }
}
```

**IMPORTANT**: Only pre-approved functions are allowed:
- formatDate
- parseJson
- stringConcat
- arrayFilter
- objectMerge
"""


def get_taskflow_variable_reference_rules() -> str:
    """Get variable reference rules for TaskFlow V2."""
    return """## Variable Reference Syntax

TaskFlow V2 uses `${}` syntax for variable references:

### Input References
- `${inputs.field_name}` - Reference input field

### Step Output References
- `${step_id.output}` - Reference entire step output
- `${step_id.output.data}` - Reference 'data' field from step output
- `${step_id.output.data.nested}` - Reference nested field

### Secret References
- `${secrets.API_KEY}` - Reference secret from vault

### Examples
```json
{
  "url": "https://api.example.com/users/${inputs.user_id}",
  "template": "Hello, ${fetch_user.output.data.name}!",
  "headers": {"Authorization": "Bearer ${secrets.API_TOKEN}"}
}
```

### Common Mistakes to Avoid
- DON'T use `:source.` prefix (that's GraphAI syntax)
- DON'T use `{{}}` (that's Jinja2 syntax)
- DO use `${}` for all variable references
"""


def get_taskflow_security_rules() -> str:
    """Get security rules for TaskFlow V2."""
    return """## Security Requirements

### 1. HTTPS Only
All URLs MUST use HTTPS protocol. HTTP is not allowed.
- Valid: `https://api.example.com/endpoint`
- Invalid: `http://api.example.com/endpoint`

### 2. No Private IP Addresses
The following are blocked for SSRF protection:
- 127.x.x.x (localhost)
- 10.x.x.x (private)
- 192.168.x.x (private)
- 169.254.x.x (link-local)
- localhost

### 3. code_js Restrictions
Only these pre-approved functions are allowed:
- formatDate: Date formatting
- parseJson: JSON parsing
- stringConcat: String concatenation
- arrayFilter: Array filtering
- objectMerge: Object merging

If you need custom logic, use `transform` step with template mode.

### 4. Path Traversal Prevention
Paths in code_js config must not contain:
- `../` (parent directory)
- `..\\` (Windows parent directory)
"""


def get_taskflow_api_rules() -> str:
    """Get available expertAgent API information for TaskFlow generation.

    Issue #350 Fix: TaskFlow workflows MUST use internal expertAgent APIs
    instead of external APIs (like api.openai.com directly).
    """
    expertagent_base = settings.EXPERTAGENT_BASE_URL or "http://localhost:8004"

    return f"""## Available ExpertAgent APIs

**CRITICAL**: You MUST use ONLY these internal APIs. DO NOT use external APIs directly.
- ❌ DO NOT use: api.openai.com, api.anthropic.com, or any external LLM API
- ✅ DO use: expertAgent internal APIs listed below

### Base URL
All API endpoints use this base URL: `{expertagent_base}`
You can use the full URL or `${{secrets.EXPERTAGENT_BASE_URL}}` for dynamic resolution.

---

### 1. Direct LLM Call (MOST IMPORTANT - Use this for all LLM tasks)
**Endpoint**: `{expertagent_base}/v1/mylllm`
**Method**: POST
**Use for**: Text generation, summarization, translation, analysis, question answering

```json
{{
  "id": "llm_process",
  "type": "api_rest",
  "config": {{
    "step_type": "api_rest",
    "method": "POST",
    "url": "{expertagent_base}/v1/mylllm",
    "headers": {{"Content-Type": "application/json"}},
    "body": "{{\"user_input\": \"${{inputs.query}}\", \"system_prompt\": \"You are a helpful assistant.\", \"model\": \"gpt-4o-mini\"}}"
  }}
}}
```

**Available models**: gpt-4o-mini, gpt-4o, gemini-1.5-flash, gemini-1.5-pro, claude-3-5-sonnet
**Response**: `{{"result": "LLM response text"}}`

---

### 2. JSON Output Agent (For structured JSON output)
**Endpoint**: `{expertagent_base}/v1/aiagent/utility/jsonoutput`
**Method**: POST
**Use for**: Structured data extraction, classification, JSON formatting

```json
{{
  "id": "json_extract",
  "type": "api_rest",
  "config": {{
    "step_type": "api_rest",
    "method": "POST",
    "url": "{expertagent_base}/v1/aiagent/utility/jsonoutput",
    "headers": {{"Content-Type": "application/json"}},
    "body": "{{\"user_input\": \"${{inputs.text}}\", \"system_prompt\": \"Extract key information and return as JSON.\"}}"
  }}
}}
```

**Response**: `{{"result": {{...}}}}` (always valid JSON)

---

### 3. Gmail APIs
**Search**: `{expertagent_base}/v1/utility/gmail/search`
```json
{{"query": "is:unread", "max_results": 10}}
```
**Response**: `{{"messages": [...], "result_count": N}}`

**Send**: `{expertagent_base}/v1/utility/gmail/send`
```json
{{"to": "email@example.com", "subject": "Subject", "body": "Message body"}}
```
**Response**: `{{"result": "...", "message_id": "..."}}`

---

### 4. Google Search APIs
**Search**: `{expertagent_base}/v1/utility/google_search`
```json
{{"queries": ["search term"], "num": 10}}
```
**Response**: `{{"search_results": [...], "status": "ok"}}`

**Search Overview**: `{expertagent_base}/v1/utility/google_search_overview`
```json
{{"queries": ["search term"], "num": 10}}
```
**Response**: `{{"result": {{"text": "ok", "result": [...]}}}}`

---

### 5. Google Drive APIs
**Upload**: `{expertagent_base}/v1/utility/drive/upload`
```json
{{"file_path": "/path/to/file", "drive_folder_url": "https://drive.google.com/..."}}
```
**Response**: `{{"status": "success", "file_id": "...", "web_view_link": "..."}}`

---

### 6. Text-to-Speech APIs
**TTS + Drive**: `{expertagent_base}/v1/utility/text_to_speech_drive`
```json
{{"text": "Text to speak", "drive_folder_url": "...", "voice": "alloy"}}
```
**Response**: `{{"file_id": "...", "web_view_link": "..."}}`
**Voices**: alloy, echo, fable, onyx, nova, shimmer

---

### Summary of API Selection
| Task Type | Use This API |
|-----------|--------------|
| LLM text generation | /v1/mylllm |
| Structured JSON output | /v1/aiagent/utility/jsonoutput |
| Email search | /v1/utility/gmail/search |
| Email send | /v1/utility/gmail/send |
| Web search | /v1/utility/google_search |
| File upload | /v1/utility/drive/upload |
| Text-to-speech | /v1/utility/text_to_speech_drive |

**REMINDER**: Never use external APIs like api.openai.com. Always use the expertAgent internal APIs listed above.
"""


# Combine all rules for convenience
TASKFLOW_RULES_FULL = "\n\n".join(get_taskflow_rules())


__all__ = [
    "get_taskflow_rules",
    "get_taskflow_structure_rules",
    "get_taskflow_step_types",
    "get_taskflow_variable_reference_rules",
    "get_taskflow_security_rules",
    "get_taskflow_api_rules",
    "TASKFLOW_RULES_FULL",
]

File: mySwiftAgentCore/src/taskflowGeneratorAgent/prompts/templates/taskflow-rules.ts
/**
 * TaskFlow Rules - Rules and schema for TaskFlow workflow generation
 *
 * Issue #364: TaskFlow generation rules based on graphAiServer format
 */

/**
 * TaskFlow generation rules for LLM prompt
 */
export const TASKFLOW_RULES = `
# TaskFlow Workflow Generation Rules

## Overview
You are a workflow generator that creates TaskFlow JSON definitions compatible with graphAiServer.
Follow these rules strictly to generate valid workflows.

## Output Format
The workflow must be a valid JSON object with the following structure:

\`\`\`json
{
  "workflow_name": "string (required, snake_case)",
  "description": "string (optional)",
  "input_schema": {
    "field_name": "type_string"
  },
  "output_schema": {
    "field_name": "type_string"
  },
  "steps": [
    {
      "id": "string (required, unique step identifier)",
      "type": "string (required, one of: api_rest, transform, code_js, llm, parallel, action)",
      "description": "string (optional)",
      "config": { ... },
      "params": { ... }
    }
  ],
  "output": {
    "result_key": "\${step_id.field_name}"
  }
}
\`\`\`

**IMPORTANT Output Mapping Syntax:**
- Use \`\${step_id.field_name}\` to reference step outputs (NO ".output" in path)
- Example: \`"\${transform_step.keyword}"\` NOT \`"\${transform_step.output.keyword}"\`

**IMPORTANT Schema Format:**
- \`input_schema\` and \`output_schema\` use SIMPLE type mapping: \`{"field_name": "type"}\`
- Valid types: "string", "number", "boolean", "array", "object"
- Do NOT use JSON Schema format (no "type": "object", "properties", "required")
- Example: \`{"query": "string", "num_results": "number"}\`

## Step Types

### 1. api_rest
REST API call step for external service integration.

**CRITICAL: \`url\` is REQUIRED for api_rest steps. Never use \`capability_id\`.**

Config (REQUIRED fields):
- \`method\`: HTTP method (GET, POST, PUT, DELETE, PATCH) - REQUIRED
- \`url\`: Full API endpoint URL - REQUIRED (e.g., "http://localhost:8004/v1/utility/google_search")
- \`headers\`: Request headers object (default: {"Content-Type": "application/json"})
- \`body\`: Request body for POST/PUT/PATCH - include directly in config
- \`timeout_ms\`: Request timeout in milliseconds (default: 30000)

**Common Capability URLs:**
- google_search: http://localhost:8004/v1/utility/google_search (POST)
- gmail_send: http://localhost:8004/v1/utility/gmail/send (POST)
- json_output_agent: http://localhost:8004/v1/aiagent/utility/jsonoutput (POST)

**API Response Field Names (IMPORTANT for transform mapping):**
- google_search returns: \`search_results\` (array of {title, link, knowledge, original_query})
- gmail_send returns: \`message_id\`, \`status\`
- json_output_agent returns: structured JSON based on system_prompt specification

Example:
\`\`\`json
{
  "id": "search_step",
  "type": "api_rest",
  "config": {
    "method": "POST",
    "url": "http://localhost:8004/v1/utility/google_search",
    "headers": {"Content-Type": "application/json"},
    "body": {"queries": ["\${inputs.query}"], "num": 3},
    "timeout_ms": 30000
  },
  "params": {}
}
\`\`\`

Params:
- Usually empty \`{}\` - put request body in \`config.body\` instead

### 2. transform
Data transformation step using template or mapping.

**IMPORTANT**: Use either \`template\` OR \`mapping\` in config. DO NOT use \`expression\`.

Config (use ONE of the following):
- \`template\`: Handlebars-like template string for transformation. Use \`{{path}}\` for variable interpolation.
  Example: \`"{\\"result\\": \\"{{input.value}}\\"}"\`
- \`mapping\`: Key-value mapping from output fields to input paths using **dot notation**.
  - For input fields: \`"input.field_name"\` (singular "input")
  - For step outputs: \`"steps.step_id.field_name"\` (with "steps." prefix)
  Example: \`{ "keyword": "input.keyword", "results": "steps.search_step.search_results" }\`

**WARNING**: \`expression\` is NOT supported for security reasons. Workflows using \`expression\` will fail validation.

Params:
- Usually empty \`{}\` for transform steps

### 3. code_js
Custom JavaScript code execution in sandboxed environment.

Config:
- \`code\`: JavaScript code string
- \`timeout\`: Execution timeout in milliseconds

Params:
- Input parameters for the code

### 4. llm
LLM (Large Language Model) call step.

Config:
- \`prompt\`: User prompt template with variable interpolation (REQUIRED). Use \`{{path}}\` for variable references.
- \`model\`: Model identifier (optional, e.g., "gpt-4o-mini")
- \`temperature\`: Generation temperature 0.0-2.0 (optional, default: 0.7)
- \`max_tokens\`: Maximum tokens to generate (optional, default: 4096)
- \`system_prompt\`: System prompt for context setting (optional)

Params:
- Input parameters for variable substitution in prompt

### 5. parallel
Parallel execution of multiple sub-steps.

Config:
- \`max_concurrency\`: Maximum concurrent executions
- \`fail_fast\`: Stop on first failure (boolean)

Params:
- \`steps\`: Array of sub-step definitions

### 6. action
Predefined action execution (e.g., sending email, notifications).

Config:
- \`action_type\`: Type of action to execute
- Action-specific configuration

Params:
- Action-specific parameters

## Variable References

**IMPORTANT: Different syntax for different step types!**

### For transform steps (mapping mode):
Use dot notation WITHOUT \`\${}\` brackets:
- \`input.field_name\`: Reference workflow input (singular "input", not "inputs")
- \`steps.step_id.field_name\`: Reference output from previous step (use "steps." prefix)

### For api_rest steps (in config.body):
Use \`\${}\` syntax:
- \`\${inputs.field_name}\`: Reference workflow input
- \`\${step_id.field_name}\`: Reference output from previous step (NO ".output" in path)
  - **IMPORTANT (Issue #396)**: stepResults stores the output object directly, so use \`\${step_id.field}\` NOT \`\${step_id.output.field}\`

### For secrets:
- \`\${secrets.VARIABLE_NAME}\`: Reference secret from MyVault (for API keys etc.)

**CRITICAL Transform Mapping Syntax:**
- For inputs: \`input.query\` (singular, no brackets)
- For step outputs: \`steps.search_step.results\` (with "steps." prefix)

## Best Practices

1. **Step IDs**: Use descriptive, snake_case step IDs (e.g., \`fetch_user_data\`, \`transform_result\`)
2. **Dependencies**: Steps are executed in order; reference only previous steps
3. **Error Handling**: Consider adding error handling for API calls
4. **Validation**: Ensure input/output schemas match actual data
5. **Security**: Never expose sensitive data in workflow definitions

## Common Patterns

### API Call with Transform (using template)
\`\`\`json
{
  "steps": [
    {
      "id": "fetch_data",
      "type": "api_rest",
      "config": { "method": "GET", "url": "https://api.example.com/data/\${inputs.id}" },
      "params": {}
    },
    {
      "id": "transform_response",
      "type": "transform",
      "config": { "template": "{\\"result\\": \\"\${fetch_data.data}\\"}" },
      "params": {}
    }
  ]
}
\`\`\`
**Note (Issue #396)**: Use \`\${fetch_data.data}\` NOT \`\${fetch_data.output.data}\`. The stepResults stores output directly.

### Transform with Mapping
\`\`\`json
{
  "steps": [
    {
      "id": "map_data",
      "type": "transform",
      "config": {
        "mapping": {
          "result": "steps.previous_step.result",
          "count": "steps.previous_step.count"
        }
      },
      "params": {}
    }
  ]
}
\`\`\`

**IMPORTANT**: In transform mapping, use \`steps.step_id.field\` (no .output), NOT \`\${step_id.output.field}\`

### LLM Processing
\`\`\`json
{
  "steps": [
    {
      "id": "generate_summary",
      "type": "llm",
      "config": {
        "prompt": "Summarize the following text: {{input.text}}",
        "model": "gpt-4o-mini",
        "temperature": 0.7,
        "system_prompt": "You are a helpful assistant."
      },
      "params": {}
    }
  ]
}
\`\`\`

## Capability vs External API Usage

When making API calls, distinguish between:

### Internal Capability (use capability URL)
For registered capabilities, construct the URL using the capability's endpoint:
\`\`\`json
{
  "id": "search_google",
  "type": "api_rest",
  "config": {
    "method": "POST",
    "url": "http://localhost:8004/v1/utility/google_search",
    "headers": { "Content-Type": "application/json" },
    "body": { "queries": ["\${inputs.search_term}"], "num": 3 },
    "timeout_ms": 30000
  },
  "params": {}
}
\`\`\`
**google_search response**: Returns \`search_results\` array. Use \`steps.search_google.search_results\` in transform mapping.

### External API (use url)
For external APIs with full URLs:
\`\`\`json
{
  "id": "fetch_weather",
  "type": "api_rest",
  "config": {
    "url": "https://api.weather.com/v1/forecast",
    "method": "GET"
  },
  "params": {}
}
\`\`\`

**Important Rules:**
- ALWAYS use \`url\` field for api_rest steps - it is REQUIRED
- NEVER use \`capability_id\` - it is not supported
- For internal capabilities, use their full URL (e.g., http://localhost:8004/v1/utility/google_search)
- For external APIs, use HTTPS URLs
- Put request body in \`config.body\`, not in \`params\`
`;

/**
 * JSON Schema for TaskFlow validation
 * Issue #396: Updated to match graphAiServer expected format (simple type mapping)
 */
export const TASKFLOW_JSON_SCHEMA = {
  $schema: 'http://json-schema.org/draft-07/schema#',
  type: 'object',
  required: ['workflow_name', 'input_schema', 'output_schema', 'steps', 'output'],
  properties: {
    workflow_name: {
      type: 'string',
      pattern: '^[a-z][a-z0-9_]*$',
    },
    description: {
      type: 'string',
    },
    // Issue #396: Simple type mapping format {field: "type"}
    input_schema: {
      type: 'object',
      additionalProperties: {
        type: 'string',
        enum: ['string', 'number', 'boolean', 'array', 'object', 'null'],
      },
    },
    output_schema: {
      type: 'object',
      additionalProperties: {
        type: 'string',
        enum: ['string', 'number', 'boolean', 'array', 'object', 'null'],
      },
    },
    steps: {
      type: 'array',
      items: {
        type: 'object',
        required: ['id', 'type', 'config'],
        properties: {
          id: { type: 'string' },
          type: {
            type: 'string',
            enum: ['api_rest', 'transform', 'code_js', 'llm', 'parallel', 'action'],
          },
          description: { type: 'string' },
          config: { type: 'object' },
          params: { type: 'object' },
        },
      },
    },
    output: {
      type: 'object',
      additionalProperties: { type: 'string' },
    },
  },
};
